syntax = "proto3";

package aod;

// ============================================================================
// Enumerations
// ============================================================================

enum TriggerType {
  TRIGGER_SOFTWARE = 0;
  TRIGGER_EXTERNAL = 1;
}

// ============================================================================
// Ping Command
// ============================================================================

// Request to ping the server (health check)
message PingRequest {
  // Empty for now - could add client timestamp later
}

// Response from ping command
message PingResponse {
  // Server timestamp in nanoseconds since epoch
  int64 timestamp_ns = 1;
}

// ============================================================================
// Initialize Command
// ============================================================================

// Initialize AWG hardware with channel amplitudes
message InitializeRequest {
  // Amplitude for each active channel in millivolts
  // Array size must match number of active channels in AWG_CHANNEL_MASK
  // Example: if mask = 0b0011 (channels 0,1), provide 2 amplitudes
  repeated int32 channel_amplitudes_mv = 1;
}

// Response from initialize command
message InitializeResponse {
  bool success = 1;
  string error_message = 2;
}

// ============================================================================
// Stop Command
// ============================================================================

// Stop AWG output and DMA
message StopRequest {
  // Empty - no parameters required
}

// Response from stop command
message StopResponse {
  bool success = 1;
  string error_message = 2;
}

// ============================================================================
// WaveformBatch Command
// ============================================================================

// Single waveform with interpolated tone parameters
message Waveform {
  // Delay in units of WAVEFORM_TIMESTEP
  // For first waveform: delay from trigger
  // For subsequent waveforms: delay from end of previous waveform
  int32 delay = 1;

  // Duration in units of WAVEFORM_TIMESTEP
  int32 duration = 2;

  // Number of tones
  int32 num_tones = 3;

  // Number of interpolation steps
  int32 num_steps = 4;

  // Interpolation time points (in units of WAVEFORM_TIMESTEP)
  // Size: num_steps
  // Defines the x-coordinates for interpolation
  repeated int32 time_steps = 5;

  // Tone parameters (flattened arrays)
  // Size for each: num_steps * num_channels * num_tones
  // Index: [step * num_channels * num_tones + channel * num_tones + tone]

  // Frequencies in Hz
  repeated float frequencies = 6;

  // Amplitudes (normalized, 0.0 to 1.0 typically)
  repeated float amplitudes = 7;

  // Offset phases (radians)
  repeated float offset_phases = 8;
}

// Batch of waveforms to execute sequentially
message WaveformBatchRequest {
  // Trigger type
  TriggerType trigger_type = 1;

  // Array of waveforms to execute sequentially
  // Each waveform has its own delay (from trigger for first, from previous end for others)
  repeated Waveform waveforms = 2;
}

// Response from waveform batch command
message WaveformBatchResponse {
  bool success = 1;
  string error_message = 2;
  int32 batch_id = 3;  // Unique ID for tracking this batch
}

// ============================================================================
// Top-Level Request/Response Wrappers
// ============================================================================

// Client sends Request messages to server
message Request {
  oneof command {
    PingRequest ping = 1;
    InitializeRequest initialize = 2;
    StopRequest stop = 3;
    WaveformBatchRequest waveform_batch = 4;
    // Future commands will be added here:
    // StartRequest start = 5;
    // etc.
  }
}

// Server sends Response messages back to client
message Response {
  oneof result {
    PingResponse ping = 1;
    InitializeResponse initialize = 2;
    StopResponse stop = 3;
    WaveformBatchResponse waveform_batch = 4;
    // Future responses will be added here:
    // ErrorResponse error = 5;
    // etc.
  }
}
